api/types/registry/registry.go

```diff
@@ -117,3 +121,143 @@ type DistributionInspect struct {
+	// endpoints.
+	Mirrors []Mirror `json:"Mirrors,omitempty"`
+}
+
+// NewRegistry returns a Registry and interprets input as a URL.
+func NewRegistry(input string) (Registry, error) {
+	reg := Registry{}
+	err := reg.URL.Parse(input)
+	return reg, err
+}
+
+// AddMirror interprets input as a URL and adds it as a new mirror.
+func (r *Registry) AddMirror(input string) error {
+	mir, err := NewMirror(input)
+	if err != nil {
+		return err
+	}
+	r.Mirrors = append(r.Mirrors, mir)
+	return nil
+}
+
+// ContainsMirror returns true if the URL of any mirror equals input.
+func (r *Registry) ContainsMirror(input string) bool {
+	for _, m := range r.Mirrors {
+		if m.URL.String() == input {
+			return true
+		}
+	}
+	return false
+}
+
+// Mirror holds information for a given registry mirror.
+type Mirror struct {
+	// The URL of the mirror.
+	URL RegURL `json:"URL,omitempty"`
+}
+
+// NewMirror returns a Registry and interprets input as a URL.
+func NewMirror(input string) (Mirror, error) {
+	mir := Mirror{}
+	err := mir.URL.Parse(input)
+	return mir, err
+}
+
+// RegURL is a wrapper for url.URL to unmarshal it from the JSON config and to
+// make it an embedded type for its users.
+type RegURL struct {
+	// rURL is a simple url.URL.  Notice it is no pointer to avoid potential
+	// null pointer dereferences.
+	rURL url.URL
+}
+
+// UnmarshalJSON unmarshals the byte array into the RegURL pointer.
+func (r *RegURL) UnmarshalJSON(b []byte) error {
+	var input string
+	if err := json.Unmarshal(b, &input); err != nil {
+		return err
+	}
+	return r.Parse(input)
+}
+
+// MarshalJSON marshals the RegURL.
+func (r *RegURL) MarshalJSON() ([]byte, error) {
+	return json.Marshal(r.String())
+}
+
+// Parse parses input as a URL.
+func (r *RegURL) Parse(input string) error {
+	input = strings.ToLower(input)
+	uri, err := url.Parse(input)
+	if err == nil {
+		r.rURL = *uri
+	} else {
+		return err
+	}
+	// default to https if no URI scheme is specified
+	if uri.Scheme == "" {
+		// we have to parse again to update all associated data
+		return r.Parse("https://" + input)
+	}
+
+	// sanity checks
+	if uri.Scheme != "http" && uri.Scheme != "https" {
+		return fmt.Errorf("invalid url: unsupported scheme %q in %q", uri.Scheme, uri)
+	}
+	if uri.Host == "" {
+		return fmt.Errorf("invalid url: unspecified hostname in %s", uri)
+	}
+	if uri.User != nil {
+		// strip password from output
+		uri.User = url.UserPassword(uri.User.Username(), "xxxxx")
+		return fmt.Errorf("invalid url: username/password not allowed in URI %q", uri)
+	}
+
+	return nil
+}
+
+// Host returns the host:port of the URL.
+func (r *RegURL) Host() string {
+	return r.rURL.Host
+}
+
+// Prefix returns the host:port/path of the URL.
+func (r *RegURL) Prefix() string {
+	return r.rURL.Host + r.rURL.Path
+}
+
+// IsOfficial returns true if the URL points to an official "docker.io" host.
+func (r *RegURL) IsOfficial() bool {
+	return r.rURL.Hostname() == "docker.io"
```

Comment:
Ouch, that would be bad to miss. Are there more hostnames to add here? As this is an API, we can't reference consts in other packages, so I believe we have to hardcode everything here :-(

Reply:
Yes, I don't like that as well (having to hard-code); hence me wondering if we actually _need_ to make exceptions. A quick test shows that these all seem to work:

```
docker pull busybox
docker pull library/busybox
docker pull docker.io/busybox
docker pull docker.io/library/busybox
docker pull index.docker.io/busybox
docker pull index.docker.io/library/busybox
docker pull registry-1.docker.io/library/busybox 
```

Interestingly, there's special handling for `docker.io` and `index.docker.io`, but not for `registry-1.docker.io`, so all show up as `busybox` in the local image cache, except for that one, which shows with its full reference:

```
REPOSITORY                             TAG                                       IMAGE ID            CREATED             SIZE
busybox                                latest                                    f9b6f7f7b9d3        37 hours ago        1.14MB
registry-1.docker.io/library/busybox   latest                                    f9b6f7f7b9d3        37 hours ago        1.14MB
````

Reply:
That is surprising. Thanks a lot for looking into this! Based on your findings, I suggest that we can consider `docker.io` and `registry-1.docker.io` to be official hosts, right?

Reply:
And index.docker.io

Reply:
Perhaps we should just check for `docker.io` and hosts ending in `.docker.io`

Reply:
> And index.docker.io

For sure, sorry for missing that!

Reply:
> Perhaps we should just check for docker.io and hosts ending in .docker.io

My initial reaction was positive, but now I started to hesitate, because this would ignore typos. Imagine something like `"Prefix": "i-am-a-typo.docker.io"`.

Reply:
This is mixing hostnames and image names. Images don't have names like `registry-1.docker.io` or `index.docker.io` and should not be considered "official". The "official" namespace is `docker.io`.

Reply:
In this case, we don't refer to image names but image references, which include the hostname.

Reply:
> This is mixing hostnames and image names. Images don't have names like registry-1.docker.io or index.docker.io and should not be considered "official". The "official" namespace is docker.io.

@stevvooe given https://github.com/moby/moby/pull/34319#discussion_r162021213: shall we change those semantics or keep them as they are now? Currently, pulling "index.docker.io/busybox" works ...

I working on this part of the PR at the moment and wonder what to do.